### DISCUSSION TITLE
The suggested company-backends setting to obtain desirable completion candidates when using `python-lsp-server`.

### DISCUSSION THREAD

**Discussion by hongyi-zhao** [company-mode/company-mode#1256] [General]
On Ubuntu 20.04.3 LTS, I'm using self-compiled git master version of Emacs with `python-lsp-server`. See the following code completion results given by `python-lsp-server`:

![image](https://user-images.githubusercontent.com/11155854/139530967-d39251d5-228d-4578-bde4-ede223db8e1f.png)

As you can see, company gives so many undesired/incorrect completion candidates in the above screenshot. According the docstring [here](https://github.com/python/cpython/blob/62bf263a775f4444d8b5d5841cc09be3bd53e933/Lib/re.py#L103-L118), only the following flags are allowed:

```
    A  ASCII       For string patterns, make \w, \W, \b, \B, \d, \D
                   match the corresponding ASCII character categories
                   (rather than the whole Unicode categories, which is the
                   default).
                   For bytes patterns, this flag is the only available
                   behaviour and needn't be specified.
    I  IGNORECASE  Perform case-insensitive matching.
    L  LOCALE      Make \w, \W, \b, \B, dependent on the current locale.
    M  MULTILINE   "^" matches the beginning of lines (after a newline)
                   as well as the string.
                   "$" matches the end of lines (before a newline) as well
                   as the end of the string.
    S  DOTALL      "." matches any character at all, including the newline.
    X  VERBOSE     Ignore whitespace and comments for nicer looking RE's.
    U  UNICODE     For compatibility only. Ignored for string patterns (it
                   is the default), and forbidden for bytes patterns.
```
So, I want to know the suggested company-backends setting to obtain desirable completion candidates when using `python-lsp-server`. Also see the issue [here](https://github.com/python-lsp/python-lsp-server/issues/101) for relevant discussion.

Regards,
HZ

**Comment by dgutov** [comment]
> According the docstring here, only the following flags are allowed:

The remaining completions are not flags. They are functions (and classes, and types) in the `re` module.

**Reply by hongyi-zhao** [reply]
So, how can I only show legitimate candidates here, or let company display the legitimate candidates in the corresponding context at the top of the candidates list?

**Reply by dgutov** [reply]
They are all legitimate (Python is dynamic). I don't think there is a completer which would limit the suggestions as you seem to want here.

You can ask, though. The authors of `python-lsp-server` would be the people to bring up this question with.

**Reply by hongyi-zhao** [reply]
See [here](https://github.com/python-lsp/python-lsp-server/issues/101#issuecomment-955839044) and [here](https://github.com/davidhalter/jedi/issues/1815) for some relevant explanation.
